File: modules/bool/app/model/Assignment.java
package model;

import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.SortedSet;
import java.util.TreeSet;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Assignment {

  private Map<Character, Boolean> assignments = new HashMap<>();

  public Assignment(char var, boolean value) {
    assignments.put(var, value);
  }

  public Assignment(Map<Character, Boolean> items) {
    // Constructor for test, used with <b>IMMUTABLE</b>Map
    assignments.putAll(items);
  }

  public static List<Assignment> generateAllAssignments(Collection<Character> variables) {
    if(variables.isEmpty())
      throw new IllegalArgumentException("Cannot generate assignments for 0 variables!");

    SortedSet<Character> distinctVars = new TreeSet<>(variables);
    return generateAllAssignments(new LinkedList<>(distinctVars));

  }

  /**
   * Generates all possible assignments for this combination of variables
   * recursively
   *
   * @param variables
   *          the variables to generate the assignments for
   * @return all assignments for this combination of variables
   */
  private static List<Assignment> generateAllAssignments(List<Character> variables) {
    char variable = variables.get(0);

    if(variables.size() == 1)
      return Arrays.asList(new Assignment(variable, false), new Assignment(variable, true));

    List<Assignment> falseAssignments = generateAllAssignments(variables.subList(1, variables.size()));
    falseAssignments.forEach(a -> a.setAssignment(variable, false));

    List<Assignment> trueAssignments = generateAllAssignments(variables.subList(1, variables.size()));
    trueAssignments.forEach(a -> a.setAssignment(variable, true));

    return Stream.concat(falseAssignments.stream(), trueAssignments.stream()).collect(Collectors.toList());
  }

  public char asChar(char variable) {
    return get(variable) ? '1' : '0';
  }

  public boolean get(char variable) {
    return (assignments.containsKey(variable) && assignments.get(variable));
  }

  public String getAssignmentsForJson() {
    return toString();
  }

  public String getColor() {
    if(get(BooleanQuestion.LEARNER_VARIABLE) == get(BooleanQuestion.SOLUTION_VARIABLE))
      return "success";
    else
      return "danger";
  }

  public char getLearnerValue() {
    return asChar(BooleanQuestion.LEARNER_VARIABLE);
  }

  public List<Character> getVariables() {
    return assignments.entrySet().stream()
        .filter(a -> a.getKey() != BooleanQuestion.SOLUTION_VARIABLE && a.getKey() != BooleanQuestion.LEARNER_VARIABLE)
        .map(Entry::getKey).collect(Collectors.toList());
  }

  public boolean isSet(char variable) {
    return assignments.get(variable) != null;
  }

  public void setAssignment(char variable, boolean value) {
    assignments.put(variable, value);
  }

  @Override
  public String toString() {
    return assignments.entrySet().stream()
        .filter(a -> a.getKey() != BooleanQuestion.SOLUTION_VARIABLE && a.getKey() != BooleanQuestion.LEARNER_VARIABLE)
        .map(entry -> entry.getKey() + ": " + entry.getValue()).collect(Collectors.joining(", "));
  }

}

File: modules/bool/app/model/Assignment.scala
package model;

import scala.collection.mutable.MapBuilder
import scala.language.implicitConversions
import model.ScalaNode._

class Assignment(assignments: Map[Variable, Boolean]) {

  val SOL_VAR = BooleanQuestion.SOLUTION_VARIABLE
  val LEA_VAR = BooleanQuestion.LEARNER_VARIABLE

  def asChar(variable: Variable) = if (apply(variable)) '1' else '0'

  def apply(variable: Variable) = assignments(variable)

  def isCorrect = apply(LEA_VAR) == apply(SOL_VAR)

  def getColor = if (isCorrect) "success" else "danger";

  def getLearnerValueAsChar = asChar(LEA_VAR)

  def variables = assignments.keys.filter(variable => variable != SOL_VAR && variable != LEA_VAR)

  def isSet(variable: Variable) = assignments.isDefinedAt(variable)

  override def toString = assignments
    .filter(as => as._1 != LEA_VAR && as._1 != SOL_VAR)
    .map(as => as._1 + ":" + (if (as._2) "1" else "0"))
    .mkString(",")

  def +(toAssign: (Variable, Boolean)) = new Assignment(assignments = assignments + (toAssign._1 -> toAssign._2))
}

object Assignment {
  implicit def intToBool(value: Int) = value != 0

  def apply(assigns: (Variable, Boolean)*) = new Assignment(assigns.toMap)

  def generateAllAssignments(variables: List[Variable]): List[Assignment] = {
    variables.sorted.reverse match {
      case Nil => List.empty
      case head :: Nil => List(Assignment(head -> false), Assignment(head -> true))
      case head :: tail => {
        val falseAssignments, trueAssignments = generateAllAssignments(tail)
        falseAssignments.map(_ + (head -> false)) ++ trueAssignments.map(_ + (head -> true))
      }
    }
  }

  def getNF(assignments: List[Assignment], takePos: Boolean, innerF: (ScalaNode, ScalaNode) => ScalaNode, outerF: (ScalaNode, ScalaNode) => ScalaNode) = assignments
    .filter(takePos ^ _(BooleanQuestion.SOLUTION_VARIABLE))
    .map(as => as.variables.map(v => if (takePos ^ as(v)) v else ScalaNode.not(v)).reduceLeft(innerF)) match {
      case Nil => takePos
      case l => l.reduceLeft(outerF)
    }
  
  def getDisjunktiveNormalForm(assignments: List[Assignment]) = getNF(assignments, false, AndScalaNode(_, _), OrScalaNode(_, _))

  def getKonjunktiveNormalForm(assignments: List[Assignment]) = getNF(assignments, true, OrScalaNode(_, _), AndScalaNode(_, _))
}
